demos/device-shadow-for-aws-iot-embedded-sdk/shadow_demo_main.c

```diff
@@ -0,0 +1,679 @@
+ * or not. If it is, it handles the message depending on the message type.
+ */
+static void prvEventCallback( MQTTContext_t * pxMqttContext,
+                              MQTTPacketInfo_t * pxPacketInfo,
+                              MQTTDeserializedInfo_t * pxDeserializedInfo )
+{
+    ShadowMessageType_t messageType = ShadowMessageTypeMaxNum;
+    const char * pcThingName = NULL;
+    uint16_t usThingNameLength = 0U;
+    uint16_t usPacketIdentifier;
+
+    ( void ) pxMqttContext;
+
+    assert( pxDeserializedInfo != NULL );
+    assert( pxMqttContext != NULL );
+    assert( pxPacketInfo != NULL );
+
+    usPacketIdentifier = pxDeserializedInfo->packetIdentifier;
+
+    /* Handle incoming publish. The lower 4 bits of the publish packet
+     * type is used for the dup, QoS, and retain flags. Hence masking
+     * out the lower bits to check if the packet is publish. */
+    if( ( pxPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+    {
+        assert( pxDeserializedInfo->pPublishInfo != NULL );
+        LogInfo( ( "pPublishInfo->pTopicName:%s.\n\n", pxDeserializedInfo->pPublishInfo->pTopicName ) );
+
+        /* Let the Device Shadow library tell us whether this is a device shadow message. */
+        if( SHADOW_SUCCESS == Shadow_MatchTopic( pxDeserializedInfo->pPublishInfo->pTopicName,
+                                                 pxDeserializedInfo->pPublishInfo->topicNameLength,
+                                                 &messageType,
+                                                 &pcThingName,
+                                                 &usThingNameLength ) )
+        {
+            /* Upon successful return, the messageType has been filled in. */
+            if( messageType == ShadowMessageTypeUpdateDelta )
+            {
+                /* Handler function to process payload. */
+                prvupdateDeltaHandler( pxDeserializedInfo->pPublishInfo );
+            }
+            else if( messageType == ShadowMessageTypeUpdateAccepted )
+            {
+                /* Handler function to process payload. */
+                prvupdateAcceptedHandler( pxDeserializedInfo->pPublishInfo );
+            }
+            else if( messageType == ShadowMessageTypeUpdateDocuments )
+            {
+                LogInfo( ( "/update/documents json payload:%s.\n\n", ( const char * ) pxDeserializedInfo->pPublishInfo->pPayload ) );
+            }
+            else if( messageType == ShadowMessageTypeUpdateRejected )
+            {
+                LogInfo( ( "/update/rejected json payload:%s.\n\n", ( const char * ) pxDeserializedInfo->pPublishInfo->pPayload ) );
+            }
+            else
+            {
+                LogInfo( ( "Other message type:%d !!\n\n", messageType ) );
+            }
+        }
+        else
+        {
+            LogError( ( "Shadow_MatchTopic parse failed:%s !!\n\n", ( const char * ) pxDeserializedInfo->pPublishInfo->pTopicName ) );
```

Comment:
Is the demo unsuccessful if Shadow_MatchTopic() fails?

Reply:
Basically yes, but we want to printout the messages to make users aware of this parse failed information.

Reply:
Then I think we should set some global variable for these callback error cases, so that RunDeviceShadowDemo can check that variable to return EXIT_FAILURE if this error occurs.